### PR TITLE
Apply new code of named collections (from #43147) to external table engines part 1

### PR DESCRIPTION
**Pull request by kssenii**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Recently a new implementation was added for named collections (https://github.com/ClickHouse/ClickHouse/pull/43147), but the code was not applied to external table engines. This is the first step to substitute old code with the new.

### CODE REVIEW

File: src/Storages/NamedCollections/NamedCollectionUtils.cpp [hunk 1]
```diff
@@ -78,7 +78,7 @@ class LoadFromConfig
         /// (`enumerate_result` == <collection_path>).
         const bool collection_is_empty = enumerate_result.size() == 1
             && *enumerate_result.begin() == collection_prefix;
-        std::set<std::string> keys;
+        std::set<std::string, std::less<>> keys;
```
Comment (nikitamikhaylov) [inline src/Storages/NamedCollections/NamedCollectionUtils.cpp:81]: Why the order is important here?

File: src/Storages/StorageS3.cpp [hunk 2]
```diff
@@ -28,6 +28,8 @@
 #include <Storages/getVirtualsForStorage.h>
 #include <Storages/checkAndGetLiteralArgument.h>
 #include <Storages/StorageURL.h>
+#include <Storages/NamedCollections/NamedCollectionsHelpers.h>
```
Comment (nikitamikhaylov) [inline src/Storages/StorageS3.cpp:31]: Why NamedCollections are in the /Storages folder? I expect them to be generic and be dependent only on AST

**Review by nikitamikhaylov — APPROVED** [review]

File: src/Storages/StorageS3.cpp (hunk 2 above)
Comment (kssenii) [reply src/Storages/StorageS3.cpp:31]: just because they are mostly used for storages, can be moved easily

**Comment by kssenii** [comment]
> Integration tests (asan) [1/6] — fail: 1

test_merge_tree_hdfs/test.py::test_attach_detach_partition

> Integration tests (release) [3/4] — fail: 1

test_merge_tree_hdfs/test.py::test_table_manipulations

> Stateless tests (debug) [4/5] — fail: 1

01051_system_stack_trace

> Stateless tests (debug) [2/5] — fail: 14

did not find any fatal message in log, but the failed tests are not related to s3, which was changed in this PR

File: src/Storages/NamedCollections/NamedCollectionUtils.cpp (hunk 1 above)
Comment (kssenii) [reply src/Storages/NamedCollections/NamedCollectionUtils.cpp:81]: I'll fix in the next PR. Do not remember for sure, but `unordered_set` should be ok too.

**Pull request merged**